homeassistant/components/caldav/calendar.py

```diff
@@ -193,27 +193,49 @@ def __init__(self, calendar, include_all_day, search):
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
         """Get the latest data."""
+        start_of_today = dt.start_of_local_day()
+        start_of_tomorrow = dt.start_of_local_day() + timedelta(days=1)
+
         # We have to retrieve the results for the whole day as the server
         # won't return events that have already started
-        results = self.calendar.date_search(
-            dt.start_of_local_day(), dt.start_of_local_day() + timedelta(days=1)
-        )
+        results = self.calendar.date_search(start_of_today, start_of_tomorrow)
+
+        # Create new vevents for each recurrence of an event that happens today.
+        # For recurring events, some servers return the original event with recurrence rules
+        # and they would not be properly parsed using their original start/end dates.
+        vevents = [event.instance.vevent for event in results]
+        new_vevents = []
+        for vevent in vevents:
+            for start_dt in vevent.getrruleset() or []:
+                _start_of_today = start_of_today
+                _start_of_tomorrow = start_of_tomorrow
+                if self.is_all_day(vevent):
+                    start_dt = start_dt.date()
+                    _start_of_today = _start_of_today.date()
+                    _start_of_tomorrow = _start_of_tomorrow.date()
+                if _start_of_today <= start_dt < _start_of_tomorrow:
+                    new_vevent = vevent
```

Comment:
Now we don't deep copy, but we do mutate a `vevent` instance. This is not good because we don't break inside this if-statement, so could technically end up in this if-statement twice, mutating the same event and putting it in the list again ?

Reply:
So revert to `deepcopy` or add unit tests?

Reply:
More unit tests added anyway. :)

Reply:
Deepcopy is an internal function and vevent is from an external package. How would deepcopy know how to copy it ? It doesn't. From the docs of deepcopy:

> It does “copy” functions and classes (shallow and deeply), by returning the original object unchanged

Reply:
I don't understand how we're not mutating the original vevent instance.

Reply:
I need to add more unit tests.

Reply:
We were mutating the original instance like you said; my unit tests were just inadequate.

Reply:
OK, `event` has a `copy()` method so I reworked it to use that. Now it works. For realsies this time.